tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,763 @@
+func (t *Tensor) setCMemAsAlreadyRelease() {
+	t.memReleased = true
+}
+
+// FreeAllocMem Method used telease the C allocated memory for this tensor.
+func (t *Tensor) FreeAllocMem() {
+	// We can't clean the tensor here in case of it had been  used as in
+	// input parameter since on tensorflow/core/client/tensor_c_api.cc the
+	// function TF_Run_Helper is cleaning the input tensors after every
+	// execution what can cause a a double free or corruption error in C++
+	// since there is no way to determine if a tensor had been previously
+	// cleaned.
+	if !t.memReleased {
+		TF_DeleteTensor(t.tensor)
+	}
+}
+
+// ErrInvalidTensorType The data type of the tensor is not compatible
+// with the expected data type on this function.
+type ErrInvalidTensorType struct {
+	tensorType   DataType
+	expectedType DataType
+}
+
+func (e *ErrInvalidTensorType) Error() string {
+	return fmt.Sprintf("Invalid tensor data type, tensor data type: '%s', required data type: '%s'", e.tensorType, e.expectedType)
+}
+
+// ErrTensorTypeNotSupported The tensor type is still not supported.
+type ErrTensorTypeNotSupported struct {
+	tensotType DataType
+}
+
+func (e *ErrTensorTypeNotSupported) Error() string {
+	return fmt.Sprintf("The tensor data type '%s' is still not supported", e.tensotType)
+}
+
+// ErrDimsOutOfTensorRange The number of specified dimensions doesn't
+// match with the tensor dimensions.
+type ErrDimsOutOfTensorRange struct {
+	tensorDim int
+	specDims  int
+}
+
+func (e *ErrDimsOutOfTensorRange) Error() string {
+	return fmt.Sprintf("The number of specified dimensions (%d) doesn't match with the tensor dimensions (%d)", e.specDims, e.tensorDim)
+}
+
+// ErrIndexOutOfRange The specified index is out of one of the dimensions range.
+type ErrIndexOutOfRange struct {
+	dim       int
+	index     int
+	dimsRange int
+}
+
+func (e *ErrIndexOutOfRange) Error() string {
+	return fmt.Sprintf("The specified index %d is out of the dimension  %d range: %d", e.index, e.dim, e.dimsRange)
+}
+
+// ErrSliceExpected The argument must be an Slice.
+type ErrSliceExpected struct {
+	dataType string
+}
+
+func (e *ErrSliceExpected) Error() string {
+	return fmt.Sprintf("The argument must be an Slice, but the data type is: '%s'", e.dataType)
+}
+
+// ErrDataTypeNotSupported The data type is still not suported.
+type ErrDataTypeNotSupported struct {
+	dataType string
+}
+
+func (e *ErrDataTypeNotSupported) Error() string {
+	return fmt.Sprintf("The type of the provided data is still not suported: '%s'", e.dataType)
+}
+
+func serialize(data interface{}, deep int, dimsIn [][]int64) (ser []interface{}, dims [][]int64, dataType DataType, dataSize int64, err error) {
+	v := reflect.ValueOf(data)
+	dims = dimsIn
+
+	if len(dims) == deep {
+		dims = append(dims, []int64{int64(v.Len())})
+	}
+	// Check the value of the elements on this slice, if they are still
+	// slices call to recursivity, if now just add the results
```

Comment:
// Check the value of the elements in this slice.  If they
// are slices, recursively serialize them, otherwise add the results.